subprojects/core/src/main/java/org/gradle/caching/internal/BuildCacheServices.java

```diff
@@ -104,9 +104,13 @@ public void registerBuildServices(ServiceRegistration registration) {
             @Provides
             BuildCacheConfigurationInternal createBuildCacheConfiguration(
                 Instantiator instantiator,
+                InstantiatorFactory instantiatorFactory,
+                ServiceRegistry services,
                 List<BuildCacheServiceRegistration> allBuildCacheServiceFactories
             ) {
-                return instantiator.newInstance(DefaultBuildCacheConfiguration.class, instantiator, allBuildCacheServiceFactories);
+                // We need to create an instantiator that has access to ObjectFactory
+                Instantiator buildScopedInstantiator = instantiatorFactory.decorate(services);
```

Comment:
💭 Here we are creating a new instantiator, since the one that gets injected to this method, `Instantiator instantiator`, doesn't have access to `ObjectFactory`.